### PR TITLE
Feat/DX-648 implemented a warning message if user is using older version of cli

### PR DESCRIPTION
**Pull request by naman-contentstack**
- Updated the CLI version 1.31.0 -> 1.32.0 
- Added a pre run hook to check if the user is using the latest version of the CLI.
- Will be checking with NPM registry if a new version is out every 3 days.
- Currently warning only once a day and it will reset at midnight.

### CODE REVIEW

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
+import { cliux, configHandler, HttpClient, LoggerService } from '@contentstack/cli-utilities';
+import * as semver from 'semver';
+interface ICacheData {
+  lastChecked: number;
+  lastWarnedDate: string;
+  latestVersion: string;
+}
+interface IVersionUpgradeWarningFrequency {
+  versionSyncDuration: number;
+}
```
Comment (aman19K) [inline packages/contentstack/src/hooks/prerun/latest-version-warning.ts:10]: @naman-contentstack , can you pls move type def in interface folder ?

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+const versionUpgradeWarningFrequency: IVersionUpgradeWarningFrequency = {
+  versionSyncDuration: 3 * 24 * 60 * 60 * 1000,
+};
+export default async function (_opts): Promise<void> {
+  const now = Date.now();
+  const today = new Date().toISOString().split('T')[0];
+  let logger!: LoggerService;
+  logger = new LoggerService(process.cwd(), 'cli-log');
```
Comment (aman19K) [inline packages/contentstack/src/hooks/prerun/latest-version-warning.ts:18]: @naman-contentstack , can we declare and initialize in same line?

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      return;
+    }
+  }
+
+  // Show warning if an update is available and last warning was > 3 hours ago
+  if (semver.gt(cache.latestVersion, this.config.version) && cache.lastWarnedDate !== today) {
+    cliux.print(
+      `You are using version ${this.config.version}, but the latest version is ${cache.latestVersion}. Please update your CLI for the best experience.`,
+      { color: 'yellow' },
+    );
```
Comment (aman19K) [inline packages/contentstack/src/hooks/prerun/latest-version-warning.ts:55]: @naman-contentstack , pls check formatting

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  if (now - cache.lastChecked > versionUpgradeWarningFrequencyConfig.versionSyncDuration) {
+    try {
+      const latestVersion = (await httpClient.get(`https://registry.npmjs.org/@contentstack/cli/latest`)).data.version;
+      cache.latestVersion = latestVersion;
+      cache.lastChecked = now;
+
+      // Save updated cache
+      configHandler.set('versionUpgradeWarningCache', cache);
+    } catch (error) {
+      logger.error('Failed to check the latest version', error);
```
Comment (aman19K) [inline packages/contentstack/src/hooks/prerun/latest-version-warning.ts:45]: @naman-contentstack , are we displaying error on console as well? If so, it is not a good idea to throw the entire api error on the console.

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts (hunk 1 above)
Comment (naman-contentstack) [reply packages/contentstack/src/hooks/prerun/latest-version-warning.ts]: Moved it to type def file, please check

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  if (configHandler.get('versionUpgradeWarningCache')) {
+    cache = configHandler.get('versionUpgradeWarningCache');
+  }
+
+  // Perform update check if needed
+  const httpClient = new HttpClient();
+
+  if (now - cache.lastChecked > versionUpgradeWarningFrequencyConfig.versionSyncDuration) {
+    try {
+      const latestVersion = (await httpClient.get(`https://registry.npmjs.org/@contentstack/cli/latest`)).data.version;
```
Comment (aman19K) [inline packages/contentstack/src/hooks/prerun/latest-version-warning.ts:32]: @naman-contentstack , what if data is empty object? We'll get undefined error. Try to use optional chaining.

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts (hunk 4 above)
Comment (naman-contentstack) [reply packages/contentstack/src/hooks/prerun/latest-version-warning.ts:39]: No we are not throwing the error on console, only in the log file

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts (hunk 3 above)
Comment (naman-contentstack) [reply packages/contentstack/src/hooks/prerun/latest-version-warning.ts:48]: Using prettier for code formatting

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts (hunk 2 above)
Comment (naman-contentstack) [reply packages/contentstack/src/hooks/prerun/latest-version-warning.ts]: Done

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts (hunk 5 above)
Comment (naman-contentstack) [reply packages/contentstack/src/hooks/prerun/latest-version-warning.ts]: Added another fallback, printing the message in log file

**Comment by aman19K** [comment]
@naman-contentstack , LGTM

**Review by aman19K — DISMISSED** [review]

**Review by netrajpatel — DISMISSED** [review]

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import * as semver from 'semver';
+import { IVersionUpgradeCache, IVersionUpgradeWarningFrequency } from '../../interfaces';
+
+const versionUpgradeWarningFrequency: IVersionUpgradeWarningFrequency = {
+  versionSyncDuration: 3 * 24 * 60 * 60 * 1000,
+};
+export default async function (_opts): Promise<void> {
+  const now = Date.now();
+  const today = new Date().toISOString().split('T')[0];
+  const logger: LoggerService = new LoggerService(process.cwd(), 'cli-log');
```
[the remark this answers is not in the document]
Comment (cs-raj) [reply packages/contentstack/src/hooks/prerun/latest-version-warning.ts:11]: Hi @naman-contentstack can you please use the following snippet here 
const logger: LoggerService = new LoggerService(process.env.CS_CLI_LOG_PATH || process.cwd(), 'cli-log');
while using cli in lambda we cannot create the file directly we need to create it in tmp

**Review by cs-raj — APPROVED** [review]

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (naman-contentstack) [reply packages/contentstack/src/hooks/prerun/latest-version-warning.ts]: Updated according to above specs

**Review by cs-raj — DISMISSED** [review]

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts (hunk 4 above)
Comment (shafeeqd959) [reply packages/contentstack/src/hooks/prerun/latest-version-warning.ts:43]: logger will display in the console ?

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      configHandler.set('versionUpgradeWarningCache', cache);
+    } catch (error) {
+      logger.error('Failed to check the latest version', error);
+      return;
+    }
+  }
+
+  // Show warning if an update is available and last warning was > 3 hours ago
+  if (semver.gt(cache.latestVersion, this.config.version) && cache.lastWarnedDate !== today) {
+    cliux.print(
```
Comment (shafeeqd959) [inline packages/contentstack/src/hooks/prerun/latest-version-warning.ts:50]: Is it 3 hours, or once in a day ?

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts (hunk 4 above)
Comment (naman-contentstack) [reply packages/contentstack/src/hooks/prerun/latest-version-warning.ts:43]: No, logger will display in the log file. specifically `cli-log.log` file

File: packages/contentstack/src/hooks/prerun/latest-version-warning.ts (hunk 7 above)
Comment (naman-contentstack) [reply packages/contentstack/src/hooks/prerun/latest-version-warning.ts:50]: It is once in a day, will update the commented code as well.

File: packages/contentstack/package.json [hunk 8]
```diff
@@ -61,6 +61,7 @@
     "@types/mkdirp": "^1.0.2",
     "@types/mocha": "^8.2.3",
     "@types/node": "^14.18.63",
+    "@types/semver": "^7.5.8",
```
Comment (shafeeqd959) [inline packages/contentstack/package.json:64]: Bump the version, this will be a minor bump

Comment (naman-contentstack) [reply packages/contentstack/package.json:64]: done

**Review by cs-raj — APPROVED** [review]

**Review by cs-raj — APPROVED** [review]

**Review by shafeeqd959 — APPROVED** [review]

**Pull request merged**